### PR TITLE
ARROW-1061: [C++] Harden decimal parsing against invalid strings

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Review by wesm — APPROVED** [review]
+1. Can you rebase?

**Comment by cpcloud** [comment]
Yep

File: cpp/src/arrow/util/decimal.cc [hunk 1]
```diff
@@ -59,25 +68,58 @@ ARROW_EXPORT Status FromString(
...
+  if (charp != end) {
+    while (charp != end && isdigit(*charp)) {
+      ++charp;
+    }
+
+    // The while loop has ended before the end of the string which means we've hit a
+    // character that isn't a base ten digit
+    if (charp != end) {
+      std::stringstream ss;
+      ss << "Found non base test digit character '" << *charp
```
Comment (cpcloud) [inline cpp/src/arrow/util/decimal.cc]: There's a typo here, `s/test/ten/`.

Comment (cpcloud) [inline cpp/src/arrow/util/decimal.cc]: Fixed.

**Comment by cpcloud** [comment]
@wesm Here's the appveyor build: https://ci.appveyor.com/project/cpcloud/arrow/build/1.0.170

**Comment by wesm** [comment]
Thanks!

**Pull request closed without merging**